src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -87,13 +128,97 @@ public override TypedExpression Transform(TypedExpression ex)
             return base.Transform(ex);
         }
 
-        private string CreateTypePrefix(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        private Concretion CreateConcretion(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        {
+            Concretion result = new Concretion();
+            foreach (var kvp in typeDict)
+            {
+                result.Add((kvp.Key.TypeName, kvp.Value.Resolution));
+            }
+            return result;
+        }
+
+        private Identifier GetConcreteIdentifier(Identifier.GlobalCallable globalCallable, ImmutableDictionary<QsTypeParameter, ResolvedType> types)
+        {
+            Concretion target = CreateConcretion(types);
+
+            // TODO: handle if globalCallable is not in Generics
+
+            var (original, concretions) = Generics[globalCallable.Item];
+
+            NonNullable<string> name;
+
+            if (concretions.ContainsKey(target))
```

Comment:
Here we may want to actually use the logic in the symbol table - we can chat about that later though (just put it on your todo list).